scalyr_agent/scalyr_client.py

```diff
@@ -1540,40 +1542,54 @@ def __set_attributes(self, thread_id, attributes, log_line_attributes=False):
 
         self.__serialization_base = tmp_buffer.getvalue()
 
-    def add_missing_attributes(self, attributes, log_line_attributes=False):
-        """ Adds items attributes to the base_event's attributes if the base_event doesn't
+    def __get_attributes_to_serialize(self):
+        if self.__disable_logfile_addevents_format:
+            result = dict()
+            if self.__parent_event:
+                result = dict(self.__parent_event.__attrs)
+            if self.__attrs:
+                result.update(self.__attrs)
+            return result
+        else:
+            return self.__attrs
+
+    def add_attributes(self, attributes, overwrite_existing=False):
+        """ Adds items attributes to __attrs if the __parent_event doesn't
         already have those attributes set.
 
-        If log_line_attributes is True the attributes will be added to the serialized form of the Event, but not
-        to its `__attrs` field, so they won't be inherited by any events created with this as a base.
+        If overwrite_existing is True an attribute will be added even if the __parent_event has the same key defined,
+        unless the value also matched to avoid wasting bytes.
         """
-        if self.__attrs:
+        if attributes:
+            attributes = dict(attributes)
+
+        if self.__parent_event:
             changed = False
-            new_attrs = dict()
-            if not log_line_attributes:
-                new_attrs = dict(self.__attrs)
             for key, value in six.iteritems(attributes):
-                if log_line_attributes or not key in self.__attrs:
+                if key not in self.__parent_event.__attrs or (
+                    overwrite_existing and self.__parent_event.__attrs[key] != value
+                ):
                     changed = True
-                    new_attrs[key] = value
+                    self.__attrs[key] = value
 
             if changed:
-                self.__set_attributes(
-                    self.__thread_id, new_attrs, log_line_attributes=log_line_attributes
-                )
+                self.__set_attributes(self.__thread_id, self.__attrs)
         else:
-            self.__set_attributes(
-                self.__thread_id, attributes, log_line_attributes=log_line_attributes
-            )
+            self.__set_attributes(self.__thread_id, attributes)
```

Comment:
In this case, we should be adding the new attributes to the ones existing in `__attrs`, but you are just overwriting `__attrs` with `attributes`.